src/assemble.test.ts

```diff
@@ -68,3 +68,96 @@ test('assemble hasMany null', () => {
     { jobNumber: 2, locations: [] },
   ]);
 });
+test('assemble hasOne', () => {
+  const objects = [{ 'p.personId': 1 }, { 'p.personId': 2 }];
+
+  const relations: Relations = {
+    person: {
+      contact: ['hasOne', 'contactDetails'],
+    },
+  };
+  const aliases = {
+    p: 'person',
+    c: 'contact',
+  };
+  const identifiers = ['p.personId', 'c.contactId'];
+
+  assert.deepStrictEqual(assemble(relations, aliases, identifiers, objects), [
+    { personId: 1 },
+    { personId: 2 },
+  ]);
+});
+
+test('assemble hasOne relationship', () => {
+  const objects = [
+    { 'p.personId': 1, 'c.contactId': 101 },
+    { 'p.personId': 2, 'c.contactId': 102 },
+  ];
+
+  const relations: Relations = {
+    person: {
+      contact: ['hasOne', 'contactDetails'],
+    },
+  };
+  const aliases = {
+    p: 'person',
+    c: 'contact',
+  };
+  const identifiers = ['p.personId', 'c.contactId'];
+
+  assert.deepStrictEqual(assemble(relations, aliases, identifiers, objects), [
+    { personId: 1, contactDetails: { contactId: 101 } },
+    { personId: 2, contactDetails: { contactId: 102 } },
+  ]);
+});
+
+test('assemble hasOne relationship with null', () => {
+  const objects = [
+    { 'p.personId': 1, 'c.contactId': 101 },
+    { 'p.personId': 2, 'c.contactId': null },
+  ];
+
+  const relations: Relations = {
+    person: {
+      contact: ['hasOne', 'contactDetails'],
+    },
+  };
+  const aliases = {
+    p: 'person',
+    c: 'contact',
+  };
+  const identifiers = ['p.personId', 'c.contactId'];
+
+  assert.deepStrictEqual(assemble(relations, aliases, identifiers, objects), [
+    { personId: 1, contactDetails: { contactId: 101 } },
+    { personId: 2 },
+  ]);
+});
+test('assemble hasOne relationship with complex data', () => {
+  const objects = [
+    { 'p.personId': 1, 'c.contactId': 101 },
+    { 'p.personId': 2, 'c.contactId': null },
+    { 'p.personId': 3, 'c.contactId': 101 },
+    { 'p.personId': 4, 'c.contactId': 103 },
+    { 'p.personId': 5 },
+  ];
+
+  const relations: Relations = {
+    person: {
+      contact: ['hasOne', 'contactDetails'],
+    },
+  };
+  const aliases = {
+    p: 'person',
+    c: 'contact',
+  };
+  const identifiers = ['p.personId', 'c.contactId'];
+
+  assert.deepStrictEqual(assemble(relations, aliases, identifiers, objects), [
+    { personId: 1 },
+    { personId: 2 },
+    { personId: 3 },
+    { personId: 4, contactDetails: { contactId: 103 } },
+    { personId: 5 },
+  ]);
+});
```

Comment:
I don't think this is correct.

1. the database will return something for person 5, see my previous comment
2. person 1 and 3 point to the same object and that is fine from a strict point of view

Reply:
fixed